sklearn/cluster/_k_means_elkan.pyx

```diff
@@ -0,0 +1,176 @@
+cdef double d(double* a, double* b, int n_features) nogil:
+    cdef double result, tmp
+    result = 0
+    cdef int i
+    for i in range(n_features):
+        tmp = (a[0] - b[0])
+        a += 1
+        b += 1
+        result += tmp * tmp
+    return sqrt(result)
+
+
+cdef update_labels_distances_inplace(
+        double* X, double* centers, double[:, :] center_distances, int[:] labels,
+        double[:, :] lower_bounds, double[:] distances, int n_samples, int
+        n_features, int n_clusters):
+    # assigns closest center to X
+    # uses triangle inequality
+    cdef double* x
+    cdef double* c
+    cdef double d_c, dist
+    cdef int c_x, j, sample
+    for sample in range(n_samples):
+        # assign first cluster center
+        c_x = 0
+        x = X + sample * n_features
+        d_c = d(x, centers, n_features)
+        lower_bounds[sample, 0] = d_c
+        for j in range(n_clusters):
+            if d_c > center_distances[c_x, j]:
+                c = centers + j * n_features
+                dist = d(x, c, n_features)
+                lower_bounds[sample, j] = dist
+                if dist < d_c:
+                    d_c = dist
+                    c_x = j
+        labels[sample] = c_x
+        distances[sample] = d_c
+
+
+def k_means_elkan(X_, int n_clusters, init, float tol=1e-4, int max_iter=30, verbose=False):
+    """Run Elkan's k-means.
+
+    Parameters
+    ----------
+    X_ : nd-array, shape (n_samples, n_features)
+
+    n_clusters : int
+        Number of clusters to find.
+
+    init : nd-array, shape (n_clusters, n_features)
+        Initial position of centers.
+
+    tol : float, default=1e-4
+        The relative increment in cluster means before declaring convergence.
+
+    max_iter : int, default=30
+        Maximum number of passes over the dataset.
+
+    verbose : bool, default=False
+        Whether to be verbose.
+
+    """
+    #initialize
+    centers_ = init
+    cdef double* centers_p = getfloatpointer(centers_)
+    cdef double* X_p = getfloatpointer(X_)
+    cdef double* x_p
+    cdef int n_samples = X_.shape[0]
+    cdef int n_features = X_.shape[1]
+    cdef int point_index, center_index, label
+    cdef float upper_bound, distance
+    cdef double[:, :] center_distances = euclidean_distances(centers_) / 2.
+    cdef double[:, :] lower_bounds = np.zeros((n_samples, n_clusters))
+    cdef double[:] distance_next_center
+    labels_ = np.empty(n_samples, dtype=np.int32)
+    cdef int[:] labels = labels_
+    upper_bounds_ = np.empty(n_samples, dtype=np.float)
+    cdef double[:] upper_bounds = upper_bounds_
+    update_labels_distances_inplace(X_p, centers_p, center_distances, labels,
+                                    lower_bounds, upper_bounds, n_samples,
+                                    n_features, n_clusters)
+    cdef np.uint8_t[:] bounds_tight = np.ones(n_samples, dtype=np.uint8)
+    cdef np.uint8_t[:] points_to_update = np.zeros(n_samples, dtype=np.uint8)
+
+    if max_iter <= 0:
```

Comment:
Yes, it raises an error.

OverflowError: can't convert negative value to unsigned int